pkg/binlog/binlog.go

```diff
@@ -0,0 +1,295 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package binlog
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/pingcap/errors"
+)
+
+// Client is the client of binlog.
+type Client struct {
+	tls        *tls.Config
+	httpClient *http.Client
+	etcdClient *clientv3.Client
+
+	// if setted, will call HookAddr to change the address of pump/drainer
+	// before accessing pump/drainer.
+	HookAddr func(addr string) (changedAddr string)
+}
+
+func (c *Client) hookAddr(addr string) string {
+	if c.HookAddr != nil {
+		return c.HookAddr(addr)
+	}
+
+	return addr
+}
+
+// NewBinlogClient create a Client.
+func NewBinlogClient(pdEndpoint []string, tlsConfig *tls.Config) (*Client, error) {
+	etcdClient, err := clientv3.New(clientv3.Config{
```

Comment:
The etcd client and http client are used in different scenarios, but they are created and closed together at the same time, can we separate them?

Reply:
And for the etcd client, can we just reuse the function here https://github.com/pingcap/tidb-operator/blob/master/pkg/pdapi/pd_control.go#L39?

Reply:
Is it OK that we refer to the current PD client and only create a new client when TLS enabled and cache the clients if TLS is disabled?